### PR TITLE
feat(headers): add AcceptCharset header

### PR DESCRIPTION
**Pull request by hugoduncan**
Adds support for the Accept-Charset header.  Encodes the charset as a string.

### CODE REVIEW

**Comment by pyfisch** [comment]
Looks good to me, although I would prefer a stronger typed charset type.

**Comment by hugoduncan** [comment]
I am happy to add an enum for the charset type, but I think we would need to agree which list of charsets to include (and have the capability of specifying a string).  Would the list in the "Preferred MIME Name" column of the [IANA rgistry](http://www.iana.org/assignments/character-sets/character-sets.xhtml) (as suggested by [rfc2616](http://www.w3.org/Protocols/rfc2616/rfc2616-sec3.html#sec3.4)) be suitable?

**Comment by seanmonstar** [comment]
I'd prefer stronger typed also, but that list is quite long. Can either keep as String for now, or try an enum with 5-10 of the most common, and then a `Ext(String)` variant.

**Comment by pyfisch** [comment]
Actually it is not that long. It can be a separate crate.

**Comment by hugoduncan** [comment]
The suggested list has 24 entries, which seems reasonable to me.  @pyfisch Would you be interested in owning such a crate (seems it would go well alongside [rust-language-tags](https://github.com/pyfisch/rust-language-tags))? @seanmonstar would you be OK adding such a dependency?

**Comment by reem** [comment]
24 entries is short enough to live in hyper imo, but I wouldn't mind another crate in hyperium to help split things up.

**Comment by seanmonstar** [comment]
24 can fit in hyper. I made mime a separate crate as it seemed useful for others.

**Comment by hugoduncan** [comment]
Just noticed https://github.com/lifthrasiir/rust-encoding.

**Comment by seanmonstar** [comment]
Oh nice. That'd be rad if we could integrate that, since we already depend on encoding (via rust-url). @lifthrasiir thoughts about a type that we could use in the headers? A label?

**Comment by lifthrasiir** [comment]
@seanmonstar lifthrasiir/rust-encoding#64 is probably relevant. I'm not sure whether this is truly desirable or not though.

**Comment by seanmonstar** [comment]
@lifthrasiir yep, that sounds exactly what would be best here...

File: src/header/shared/charset.rs [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+            "ISO-8859-2" => Iso_8859_2,
+            "ISO-8859-3" => Iso_8859_3,
+            "ISO-8859-4" => Iso_8859_4,
+            "ISO-8859-5" => Iso_8859_5,
+            "ISO-8859-6" => Iso_8859_6,
+            "ISO-8859-7" => Iso_8859_7,
+            "ISO-8859-8" => Iso_8859_8,
+            "ISO-8859-9" => Iso_8859_9,
+            "ISO-8859-10" => Iso_8859_10,
+            "Shift-JIS" => Shift_Jis,
```
Comment (SBSTP) [inline src/header/shared/charset.rs]: Shouldn't that be "SHIFT-JIS" since you're using uppercase?

**Comment by seanmonstar** [comment]
merged manually

**Comment by SBSTP** [comment]
I'm pretty sure the `Shift-JIS` bug is still there.
I added a test, `assert_eq!(Shift_Jis,"Shift-JIS".parse().unwrap());` and it doesn't work.
The line `"Shift-JIS" => Shift_Jis,` never matches and it goes straight to `Ext("SHIFT-JIS")`.

**Comment by seanmonstar** [comment]
@SBSTP oh! i see what you mean, since `to_ascii_uppercase()` is called just before it. I was quite confused at first, since it's `Shift-JIS` in both `from_str` and `name`.

**Comment by SBSTP** [comment]
Do you want me to PR a fix?

**Comment by seanmonstar** [comment]
Sure. Though, if the name is supposed to mixed case like that, then the fix
must be about using to_uppercase. If it's not supposed to, then your
suggestion works perfectly.

**Comment by SBSTP** [comment]
I was only going to fix the parsing part, since uppercase should handle mixed case.

**Pull request closed without merging**